pypeit/flatfield.py

```diff
@@ -262,16 +306,61 @@ def fit2illumflat(self, slits, frametype='illum', initial=False, flexure_shift=N
                 continue
             # Skip those without a bspline
             # DO it
-            _slitid_img = slits.slit_img(slitidx=slit_idx, initial=initial, flexure=flexure_shift)
+            _slitid_img = slits.slit_img(slitidx=slit_idx, initial=initial, flexure=spat_flexure)
             onslit = _slitid_img == slits.spat_id[slit_idx]
             spat_coo = slits.spatial_coordinate_image(slitidx=slit_idx,
                                                       initial=initial,
                                                       slitid_img=_slitid_img,
-                                                      flexure_shift=flexure_shift)
+                                                      flexure_shift=spat_flexure)
             illumflat[onslit] = spat_bsplines[slit_idx].value(spat_coo[onslit])[0]
         # TODO -- Update the internal one?  Or remove it altogether??
         return illumflat
 
+    def fit2illumflat_fine(self, slits, tilts=None, initial=False, spat_flexure=None, frametype='illum'):
+        """
+        Main algorithm to build an image of the fine correction to the spatial illumination profile.
+
+        Only applied to good slits, which means any non-flagged or flagged
+         in the exclude_for_reducing list
+
+        Args:
+            slits (:class:`pypeit.slittrace.SlitTraceSet`):
+                Definition of the slits.
+            tilts (`numpy.ndarray`_, optional):
+                Wavelength tilts image. Tilts are needed in case a sky flat was
+                used to generate the illumflat spatial profile.
+            initial (bool, optional):
+                If True, the initial slit edges will be used.
+            spat_flexure (float, optional):
+                Spatial flexure correction in pixels.
+            frametype (str, optional):
+                The frame type should be 'illum' to return the illumflat version,
+                or 'pixel' to return the pixelflat version.
+
+        Returns:
+            `numpy.ndarray`_: The fine correction to the spatial illumination profile.
+        """
+        # Check spatial flexure type
+        if (spat_flexure is not None) and (not isinstance(spat_flexure, float)):
+            msgs.error("Spatial flexure must be None or float")
+        # Initialise the illumflat fine correction
+        flat_finecorr = np.ones(self.shape())
+        # Get the slit properties
+        _slitid_img = slits.slit_img(flexure=spat_flexure, initial=initial)
+        left, right, _ = slits.select_edges(initial=initial, flexure=spat_flexure)
+        # Loop through all slits
+        for slit_idx, slit_spat in enumerate(slits.spat_id):
+            onslit_tweak = _slitid_img == slit_spat
+            ww = np.where(onslit_tweak)
+            cut = (ww[0], ww[1])
+            xpos = (cut[1] - left[cut[0],slit_idx]) / (right[cut[0],slit_idx] - left[cut[0],slit_idx])
```

Comment:
It seems some state variable should be set that indicates whether tilts were used in the fit or not. In other words, I can imagine a situation where tilts were not used in the fit, but then accidentally passed in and then the code is going to use the tilts even though the fit was not done with the fits.